src/cuda/graph_ops.cu

```diff
@@ -1,77 +1,133 @@
+}
 
-  using it = thrust::counting_iterator<IdType>;
-  thrust::for_each(
-      thrust::device, it(0), it(size),
-      [in = column_ids.data_ptr<IdType>(),
-       in_indptr = indptr.data_ptr<IdType>(),
-       out = thrust::raw_pointer_cast(item_prefix)] __device__(int i) mutable {
-        IdType begin = in_indptr[in[i]];
-        IdType end = in_indptr[in[i] + 1];
-        out[i] = end - begin;
-      });
+template <typename IdType, typename DType>
+__global__ void GroupNormL2(IdType* indptr, IdType* e_ids, DType* data,
+                            DType* out_data, int64_t size) {
+  int64_t row = blockIdx.x * blockDim.y + threadIdx.y;
 
-  cub_exclusiveSum<IdType>(thrust::raw_pointer_cast(item_prefix), size + 1);
-  return new_indptr;
+  while (row < size) {
+    IdType start = indptr[row];
+    IdType n_edges = indptr[row + 1] - start;
+    for (int idx = threadIdx.x; idx < n_edges; idx += blockDim.x) {
+      IdType pos = (e_ids == nullptr) ? start + idx : e_ids[start + idx];
+      AtomicAdd(&out_data[row], (data == nullptr) ? 1.0 : powf(data[pos], 2));
+    }
+    row += gridDim.x * blockDim.y;
+  }
 }
 
-template <typename IdType>
-__global__ void _GetSubIndicesKernel(IdType* out_indices, IdType* indptr,
-                                     IdType* indices, IdType* sub_indptr,
-                                     IdType* column_ids, int64_t size) {
+template <typename IdType, typename DType>
+__global__ void GroupDiv(IdType* indptr, IdType* e_ids, DType* data,
+                         DType* divisor, DType* out_data, int64_t size) {
   int64_t row = blockIdx.x * blockDim.y + threadIdx.y;
 
   while (row < size) {
-    int64_t in_start = indptr[column_ids[row]];
-    int64_t out_start = sub_indptr[row];
-    int64_t n_edges = sub_indptr[row + 1] - sub_indptr[row];
+    IdType start = indptr[row];
+    IdType n_edges = indptr[row + 1] - start;
     for (int idx = threadIdx.x; idx < n_edges; idx += blockDim.x) {
-      out_indices[out_start + idx] = indices[in_start + idx];
+      IdType pos = (e_ids == nullptr) ? start + idx : e_ids[start + idx];
+      out_data[pos] = ((data == nullptr) ? 1.0 : data[pos]) / divisor[row];
     }
     row += gridDim.x * blockDim.y;
   }
 }
 
-template <typename IdType>
-torch::Tensor GetSubIndices(torch::Tensor indptr, torch::Tensor indices,
-                            torch::Tensor sub_indptr,
-                            torch::Tensor column_ids) {
-  int64_t size = column_ids.numel();
-  thrust::device_ptr<IdType> item_prefix(
-      static_cast<IdType*>(sub_indptr.data_ptr<IdType>()));
-  int n_edges = item_prefix[size];  // cpu
-  auto sub_indices = torch::zeros(n_edges, indices.options());
+torch::Tensor GraphSumCUDA(torch::Tensor indptr,
+                           torch::optional<torch::Tensor> e_ids,
+                           torch::optional<torch::Tensor> data) {
+  auto size = indptr.numel() - 1;
+  auto data_ptr =
+      (data.has_value()) ? data.value().data_ptr<_Float32>() : nullptr;
```

Comment:
Why use `_Float32` here?

Reply:
Cause the data tensor is of `torch::dtype(torch::kFloat32)`, any concerns here?

Reply:
I think template `FloatType` is better.

Reply:
> Cause the data tensor is of `torch::dtype(torch::kFloat32)`, any concerns here?

I just feel weird to have the type name of `_Float32` here. Is it from Pytorch? Why does it not have a Pytorch namespace?